### PR TITLE
Polishing up Flo integration code and improving test coverage

### PR DESCRIPTION
**Pull request by aclement**
- rewritten graph-to-text to a new algorithm that fits the
  new fan-in/fan-out model much better
- polished text-to-graph
- improved test coverage for text-to-graph and graph-to-text
- removed unnecessary code

### CODE REVIEW

File: ui/src/app/shared/services/parser.spec.ts [hunk 1]
```diff
@@ -1,508 +1,473 @@
...
+        expect(node.group).toEqual('bbb');
+        expect(node.options.get('bb')).toEqual('cc');
+    });
+
+    it('dotted arg names in stream', () => {
+        parseResult = Parser.parse('aaa --bbb.ccc.ddd=eee', 'stream');
+        expect(parseResult.lines.length).toEqual(1);
+        node = parseResult.lines[0].nodes[0];
+        expect(node.name).toEqual('aaa');
+        Array.from(node.options.keys()).forEach((k)=>console.log('key is '+k));
```
Comment (cppwfs) [inline ui/src/app/shared/services/parser.spec.ts]: Probably can remove this console.log.  (Will do on Merge).

File: ui/src/app/shared/services/parser.ts [hunk 2]
```diff
@@ -557,15 +529,16 @@ class InternalParser {
                     }
                 } else {
                     const streamdef = this.eatStream(lineNumber);
-                    console.log('JSParse: parsed to stream definition: ' + JSON.stringify(streamdef));
                     // streamDef = {"apps":[{"name":"time","start":0,"end":4},{"name":"log","start":7,"end":10}]}
```
Comment (cppwfs) [inline ui/src/app/shared/services/parser.ts]: Can we remove these commented lines.  If so, I can do it on merge.

File: ui/src/app/shared/services/parser.spec.ts (hunk 1 above)
Comment (aclement) [reply ui/src/app/shared/services/parser.spec.ts]: yep, good catch.

File: ui/src/app/shared/services/parser.ts (hunk 2 above)
Comment (aclement) [reply ui/src/app/shared/services/parser.ts]: ok

File: ui/src/app/shared/services/parser.ts [hunk 3]
```diff
@@ -253,19 +241,13 @@ class InternalParser {
     }
 
     // argValue: identifier | literal_string
```
[the remark this answers is not in the document]
Comment (aclement) [reply ui/src/app/shared/services/parser.ts]: Do we have to remove these? They embody the expected grammar rules - do you want them as jsdoc instead?

File: ui/src/app/shared/services/parser.ts [hunk 4]
```diff
@@ -297,12 +279,8 @@ class InternalParser {
     }
 
     // appArguments : DOUBLE_MINUS identifier(name) EQUALS identifier(value)
```
[the remark this answers is not in the document]
Comment (aclement) [reply ui/src/app/shared/services/parser.ts]: Do we have to remove these? They embody the expected grammar rules - do you want them as jsdoc instead?

File: ui/src/app/shared/services/parser.ts [hunk 5]
```diff
@@ -297,12 +279,8 @@ class InternalParser {
...
-    private maybeEatAppArgs() {
-        let args = null;
-        // TODO not entirely sure this first problem can happen since dashes can now be in the app names
-        if (this.peekToken(TokenKind.DOUBLE_MINUS) && this.isNextTokenAdjacent()) {
-            throw {'msg': 'Expected whitespace after app name before option', 'start': this.peekAtToken().start};
-        }
+    private maybeEatAppArgs(): Parser.Option[] {
+        let args: Parser.Option[] = null;
```
[the remark this answers is not in the document]
Comment (aclement) [reply ui/src/app/shared/services/parser.ts]: I haven't dug through it too deeply to refresh my memory but I believe we keep it null unless we discover there is at least one argument, in which case we initialize it and start capturing them. I believe we treat null differently to empty array in other places.

File: ui/src/app/shared/services/parser.ts [hunk 6]
```diff
@@ -381,10 +357,7 @@ class InternalParser {
...
             return token.data;
         }
-        if (token.kind) {
-            return token.kind;
-        }
-        return JSON.stringify(token);
+        return token.kind;
     }
 
     // (name =)
```
[the remark this answers is not in the document]
Comment (aclement) [reply ui/src/app/shared/services/parser.ts]: again these are grammar rules (my variant of BNF) intended to indicate what the attached parse function will do.

File: ui/src/app/shared/services/parser.ts [hunk 7]
```diff
@@ -494,24 +466,25 @@ class InternalParser {
...
+                    console.log('JSParse: processing ' + this.text);
+                }
                 this.tokenStream = tokenize(this.text);
-                console.log('JSParse: tokenized to ' + JSON.stringify(this.tokenStream));
+                if (InternalParser.DEBUG) {
+                    console.log('JSParse: tokenized to ' + JSON.stringify(this.tokenStream));
+                }
                 this.tokenStreamPointer = 0;
                 this.tokenStreamLength = this.tokenStream.length;
                 // time | log
```
[the remark this answers is not in the document]
Comment (aclement) [reply ui/src/app/shared/services/parser.ts]: i'll remove it but i do find these helpful when debugging as a reminder for what should be in the datastructures flying around.

File: ui/src/app/streams/flo/graph-to-text.spec.ts [hunk 8]
```diff
@@ -34,32 +36,431 @@ describe('graph-to-text', () => {
...
+        }
+    });
+
+    it('basic - one simple stream', () => {
+        createLink(createSource('time'), createSink('log'));
+        dsl = convertGraphToText(graph);
+        expect(dsl).toEqual('time | log');
+    });
+
+    it('verify direct call without dsl', () => {
```
[the remark this answers is not in the document]
Comment (aclement) [reply ui/src/app/streams/flo/graph-to-text.spec.ts]: Good catch, it is left over from when these were not quite the same - there were two convert graph functions - one provided details (including the dsl) and one just provided the dsl. Towards the end of development of the PR I removed the detailed one, which then caused two similar tests to be left behind. Deleting.

File: ui/src/app/streams/flo/graph-to-text.spec.ts [hunk 9]
```diff
@@ -34,32 +36,431 @@ describe('graph-to-text', () => {
...
+    });
+
+    it('basic - link editing in process (invalid target id)', () => {
+        const source = createSource('time');
+        const linkParams: Shapes.LinkCreationParams = {
+            source: {'id': source.id, 'port': 'output', 'selector': '.output-port'},
+            target: {'id': 'not-exist', 'port': 'input', 'selector': '.input-port'}
+        };
+        const link = Shapes.Factory.createLink(linkParams);
+        console.log('wible');
```
[the remark this answers is not in the document]
Comment (aclement) [reply ui/src/app/streams/flo/graph-to-text.spec.ts]: that's funny. I use wibble to help me find my stuff in console output but I mispelled it a couple of times. I did a search for wibble before producing the PR but not the mispelling ;)

File: ui/src/app/streams/flo/graph-to-text.spec.ts [hunk 10]
```diff
@@ -34,32 +36,431 @@ describe('graph-to-text', () => {
...
+        createLink(createSource('file'), d);
+        dsl = convertGraphToText(graph);
+        expect(dsl).toEqual('time > :d\nfile > :d');
+    });
+
+    // The next tests are building up a graph over a number of steps (each
+    // builds on the last). The final graph should be:
+    // a time source connected to four things. A cassandra sink, tapped to a jdbc
+    // sink, tapped to a websocket sink and tapped to a time-log destination. The
+    // time-log destinatino is tapped to a log sink and attached to a throughput sink.
```
[the remark this answers is not in the document]
Comment (aclement) [reply ui/src/app/streams/flo/graph-to-text.spec.ts]: My vscode editor is missing a Glenn quality spell checker.

File: ui/src/app/shared/services/parser.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (cppwfs) [reply ui/src/app/shared/services/parser.ts]: No.  these can stay.  I was being too strict.

File: ui/src/app/shared/services/parser.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (cppwfs) [reply ui/src/app/shared/services/parser.ts]: you are correct.   I missed the fact that a token (app) may not have a --arg.

**Comment by cppwfs** [comment]
LGTM

**Comment by cppwfs** [comment]
Squashed Merged

**Pull request closed without merging**